### ISSUE TITLE
Issues with large payload reserves

### ISSUE THREAD

**Issue by davidhefley** [xobotyi/beansclient#2]
When I ->reserve(), I expect to get the next job; however, when my payloads get large (~16300 characters), the reservation just hangs and nothing happens. If I kill the script, and restart it, the same thing happens.

- OS: Ubuntu 16.04.6 LTS
- PHP:  PHP 7.1.27-1+ubuntu16.04.1+deb.sury.org+1
- BeansClient: Latest xoboty/beansclient

I'm running beanstalkd via a docker container:  
     schickling/beanstalkd

I've downloaded and installed the related beanstalkd-console docker, and it seems to pull the payload OK, and if I use the command line tools (beanstalk-cli) then it also can pull the payload.

Here is a script I use to replicate

```
<?php
if (PHP_SAPI !== 'cli') die('NO');
require __DIR__ . '/vendor/autoload.php';
use xobotyi\beansclient\BeansClient;
use xobotyi\beansclient\Connection;

$str = "a";
for ($i=0;$i<15280;$i++) {
    $str.='a'; //lets build a really long string
}

$connection = new Connection('10.15.21.49', '11300', 2, false);
$beansClient = new BeansClient($connection);

$loop = 0;
$s = $beansClient->stats();

print_r($s);

while ($loop < 1500 ) {
    $loop++;
    $str .= 'a'; //slowly increase the size of the string
    echo strlen($str);
    echo "\n";
    echo "Putting job\n";
    $beansClient->useTube('default')->put($str);
    echo "Getting job\n";
    $job = $beansClient->watchTube('default')->reserve();
    $id = $job->id;
    if (empty($id)) echo "No jobs...and timedout\n";
    else echo "Job is is " . $id . "\n";
    if ($id) {
        echo "\n";
        $len = strlen($job->payload);
        echo "\n";
        echo $len;
        echo "\n";
        $beansClient->delete($id);
    } else {
        die('ooops');
    }
}
```
### I've checked
- [X] There is no similar issue from other users
- [X] Issue isn't fixed in `dev` branch

**Comment by davidhefley** [comment]
FWIW: It seems the "size" is a bit sporadic as it can very quite a bit - sometimes, all I have to do is stop my script, and re-run it, and then everything works fine...

**Comment by davidhefley** [comment]
If I terminate the beanstalk daemon during the hang-up, I get this (not sure if it helps):

PHP Fatal error:  Uncaught xobotyi\beansclient\Exception\Client: Expected CRLF[\r\n] after 1 byte(s) of data, got  in /var/www/html/MMailSvc/vendor/xobotyi/beansclient/src/BeansClient.php:108
Stack trace:
#0 /var/www/html/MMailSvc/vendor/xobotyi/beansclient/src/BeansClient.php(304): xobotyi\beansclient\BeansClient->dispatchCommand(Object(xobotyi\beansclient\Command\Reserve))
#1 /var/www/html/MMailSvc/src/massmail/server.php(214): xobotyi\beansclient\BeansClient->reserve()
#2 /var/www/html/MMailSvc/worker.php(209): ndemassmail\NDEMMSERVER->processQueuedOriginal('1', true)
#3 /var/www/html/MMailSvc/worker.php(160): processOriginals(25, Array, true)
#4 {main}
  thrown in /var/www/html/MMailSvc/vendor/xobotyi/beansclient/src/BeansClient.php on line 108

**Comment by davidhefley** [comment]
I'm wondering if it's related to the issue noted here https://www.php.net/manual/en/function.feof.php (under the warning).  It appears that beanstalkd is never sending an feof.  I changed the read function so it checks for 5 consecutive 0 length data returns, and then proceeds:

 ```
public function read(int $length) :string {
        if (!$this->socket) {
            throw new Exception\Connection(0, "Unable to read from closed connection");
        }

        $str  = '';
        $read = 0;
        $nodatacount = 0;
        while ($read < $length && !$this->feof($this->socket)) {
            $data = $this->fread($this->socket, $length);

            if ($data === false) {
                throw new Exception\Socket(sprintf("Failed to read data from socket ({$this->host}:{$this->port})"));
            }
            $datalength = strlen($data);
            if ( $datalength==0 ) $nodatacount++;
            $read += strlen($data);
            $str  .= $data;
            if ( $nodatacount == 5 ) {
                $str.="\r\n";
                break;
            }
        }

        return $str;
    }

```

**Comment by xobotyi** [comment]
@davidhefley I've resulted with the same thing.
I'll rework read functions when I'll have some time.

**Comment by xobotyi** [comment]
@davidhefley hey!
Released v1.0.1 - it uses new socket methods - possibly will solve the problem, recheck pls.

**Comment by davidhefley** [comment]
The test script seems to run perfect now. Thanks!
